go-controller/pkg/ovn/master.go

```diff
@@ -1056,3 +1061,128 @@ func (oc *Controller) deleteNodeChassis(nodeName string) error {
 
 	return nil
 }
+
+// ensureLocalRaftServerID is used to ensure there is no stale member in the Raft cluster with our address
+func (oc *Controller) ensureLocalRaftServerID(db string) {
+	var dbName string
+	var appCtl func(args ...string) (string, string, error)
+
+	ticker := time.NewTicker(60 * time.Second)
+
+	if strings.Contains(db, "ovnnb") {
+		dbName = "OVN_Northbound"
+		appCtl = util.RunOVNNBAppCtl
+	} else {
+		dbName = "OVN_Southbound"
+		appCtl = util.RunOVNSBAppCtl
+	}
+
+	klog.Infof("Starting ensure routine for local Raft Server ID on db: %s", db)
+	for {
+		select {
+		case <-ticker.C:
+			out, stderr, err := util.RunOVSDBTool("db-sid", db)
+			if err != nil {
+				klog.Warningf("Unable to get db server ID for: %s, stderr: %v, err: %v", db, stderr, err)
+				continue
+			}
+			if len(out) < 4 {
+				klog.Errorf("Invalid db id found: %s for db: %s", out, db)
+				continue
+			}
+			// server ID in raft membership is only first 4 char prefix
+			serverID := out[:4]
+			out, stderr, err = appCtl("cluster/status", dbName)
+			if err != nil {
+				klog.Warningf("Unable to get cluster status for: %s, stderr: %v, err: %v", db, stderr, err)
+				continue
+			}
+			r, _ := regexp.Compile(`Address: *((ssl|tcp):[?[a-z0-9.:]+]?)`)
+			matches := r.FindStringSubmatch(out)
+			if len(matches) < 2 {
+				klog.Warningf("Unable to parse Address for db: %s, output: %s", db, out)
+				continue
+			}
+			addr := matches[1]
+			// look for current servers in raft cluster with the same address
+			r, _ = regexp.Compile("([a-z0-9]{4}) at " + addr)
+			members := r.FindAllStringSubmatch(out, -1)
+			for _, member := range members {
+				if len(member) < 2 {
+					klog.Warningf("Unable to find server id submatch in %s", member)
+					continue
+				}
+				if member[1] != serverID {
+					// stale entry found for this node with same adddress, need to kick
+					klog.Infof("Previous stale member found: %s...kicking", member[1])
+					_, stderr, err = appCtl("cluster/kick", dbName, member[1])
+					if err != nil {
+						klog.Errorf("Error while kicking old Raft member: %s, for address: %s in db: %s,"+
+							"stderr: %v, error: %v", member[1], addr, db, stderr, err)
+					}
+				}
+			}
+		case <-oc.stopChan:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+// ensureClusterRaftMembership ensures there are no unknown members in the current Raft cluster
+func (oc *Controller) ensureClusterRaftMembership(db string) {
+	var knownMembers []string
+	var dbName string
+	var appCtl func(args ...string) (string, string, error)
+
+	ticker := time.NewTicker(60 * time.Second)
+
+	if strings.Contains(db, "ovnnb") {
+		dbName = "OVN_Northbound"
+		appCtl = util.RunOVNNBAppCtl
+		knownMembers = strings.Split(config.OvnNorth.Address, ",")
```

Comment:
This means that in order to add a new member, the configuration needs to be updated and the leader (or potentially, all the masters) needs to be restarted. right? Otherwise, as soon as it sees it, it will kick it out.

Reply:
Right. So in OCP, CNO will refresh the daemonset with the new addresses, and in CNO the ovnkube-master DS contains both ovnkube-master and the dbs. So when DS gets refreshed, everything will restart and I don't think this will be an issue.

But for upstream and to be extra safe...If somehow the new one came up and was kicked because the old members were not updated yet, then we should expect the affected db to exit and rejoin the cluster when he restarts. However I do not see any check in the db code today to ensure he is still part of the cluster. I'll look at adding that and report back.

Reply:
I think I resolved this by adding commit "Add health check to ensure Raft membership"